File: _complete_archive/_archive_2026-02-26_18-50-32/stacks/go/base/code/service_layer.tpl.go
// File: service_layer.tpl.go
// Purpose: Service layer pattern with interfaces
// Generated for: {{PROJECT_NAME}}

package services

import (
	"context"
	"errors"
	"time"
)

// Common errors
var (
	ErrNotFound      = errors.New("entity not found")
	ErrAlreadyExists = errors.New("entity already exists")
	ErrInvalidInput  = errors.New("invalid input")
	ErrUnauthorized  = errors.New("unauthorized")
)

// Result represents a service operation result
type Result[T any] struct {
	Success   bool
	Data      T
	Error     error
	ErrorCode string
}

// Ok creates a successful result
func Ok[T any](data T) Result[T] {
	return Result[T]{Success: true, Data: data}
}

// Fail creates a failed result
func Fail[T any](err error, code string) Result[T] {
	var zero T
	return Result[T]{Success: false, Error: err, ErrorCode: code, Data: zero}
}

// Repository interface
type Repository[T any, ID any] interface {
	FindByID(ctx context.Context, id ID) (*T, error)
	FindAll(ctx context.Context, opts FindOptions) ([]T, int64, error)
	Create(ctx context.Context, entity *T) error
	Update(ctx context.Context, entity *T) error
	Delete(ctx context.Context, id ID) error
}

// FindOptions for pagination and filtering
type FindOptions struct {
	Page    int
	PerPage int
	OrderBy string
	Order   string // asc, desc
	Filters map[string]interface{}
}

// DefaultFindOptions returns default options
func DefaultFindOptions() FindOptions {
	return FindOptions{
		Page:    1,
		PerPage: 20,
		OrderBy: "created_at",
		Order:   "desc",
		Filters: make(map[string]interface{}),
	}
}

// PaginatedResult for list operations
type PaginatedResult[T any] struct {
	Data       []T   `json:"data"`
	Page       int   `json:"page"`
	PerPage    int   `json:"per_page"`
	Total      int64 `json:"total"`
	TotalPages int   `json:"total_pages"`
}

// BaseService with common CRUD operations
type BaseService[T any, ID any] struct {
	repo Repository[T, ID]
}

// NewBaseService creates a new base service
func NewBaseService[T any, ID any](repo Repository[T, ID]) *BaseService[T, ID] {
	return &BaseService[T, ID]{repo: repo}
}

// Get retrieves an entity by ID
func (s *BaseService[T, ID]) Get(ctx context.Context, id ID) Result[*T] {
	entity, err := s.repo.FindByID(ctx, id)
	if err != nil {
		return Fail[*T](err, "NOT_FOUND")
	}
	return Ok(entity)
}

// List retrieves entities with pagination
func (s *BaseService[T, ID]) List(ctx context.Context, opts FindOptions) Result[PaginatedResult[T]] {
	data, total, err := s.repo.FindAll(ctx, opts)
	if err != nil {
		return Fail[PaginatedResult[T]](err, "LIST_FAILED")
	}

	totalPages := int(total) / opts.PerPage
	if int(total)%opts.PerPage > 0 {
		totalPages++
	}

	return Ok(PaginatedResult[T]{
		Data:       data,
		Page:       opts.Page,
		PerPage:    opts.PerPage,
		Total:      total,
		TotalPages: totalPages,
	})
}

// Create creates a new entity
func (s *BaseService[T, ID]) Create(ctx context.Context, entity *T) Result[*T] {
	if err := s.repo.Create(ctx, entity); err != nil {
		return Fail[*T](err, "CREATE_FAILED")
	}
	return Ok(entity)
}

// Update updates an entity
func (s *BaseService[T, ID]) Update(ctx context.Context, entity *T) Result[*T] {
	if err := s.repo.Update(ctx, entity); err != nil {
		return Fail[*T](err, "UPDATE_FAILED")
	}
	return Ok(entity)
}

// Delete deletes an entity
func (s *BaseService[T, ID]) Delete(ctx context.Context, id ID) Result[bool] {
	if err := s.repo.Delete(ctx, id); err != nil {
		return Fail[bool](err, "DELETE_FAILED")
	}
	return Ok(true)
}

// EventBus for domain events
type Event struct {
	Type        string
	AggregateID interface{}
	Payload     interface{}
	OccurredAt  time.Time
}

type EventHandler func(ctx context.Context, event Event) error

type EventBus struct {
	handlers map[string][]EventHandler
}

func NewEventBus() *EventBus {
	return &EventBus{
		handlers: make(map[string][]EventHandler),
	}
}

func (eb *EventBus) Subscribe(eventType string, handler EventHandler) {
	eb.handlers[eventType] = append(eb.handlers[eventType], handler)
}

func (eb *EventBus) Publish(ctx context.Context, event Event) error {
	handlers := eb.handlers[event.Type]
	for _, handler := range handlers {
		if err := handler(ctx, event); err != nil {
			// Log error but continue with other handlers
			continue
		}
	}
	return nil
}

// UnitOfWork for transaction management
type UnitOfWork interface {
	Begin(ctx context.Context) (context.Context, error)
	Commit(ctx context.Context) error
	Rollback(ctx context.Context) error
}

// Example User service
type UserService struct {
	*BaseService[User, uint]
	eventBus       *EventBus
	passwordHasher PasswordHasher
}

type User struct {
	ID           uint
	Email        string
	Username     string
	PasswordHash string
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

type PasswordHasher interface {
	Hash(password string) (string, error)
	Verify(password, hash string) bool
}

type UserRepository interface {
	Repository[User, uint]
	FindByEmail(ctx context.Context, email string) (*User, error)
	FindByUsername(ctx context.Context, username string) (*User, error)
	UpdateLastLogin(ctx context.Context, id uint) error
}

func NewUserService(repo UserRepository, eventBus *EventBus, hasher PasswordHasher) *UserService {
	return &UserService{
		BaseService:    NewBaseService[User, uint](repo),
		eventBus:       eventBus,
		passwordHasher: hasher,
	}
}

func (s *UserService) Register(ctx context.Context, email, password, username string) Result[*User] {
	userRepo := s.repo.(UserRepository)

	// Check if email exists
	existing, _ := userRepo.FindByEmail(ctx, email)
	if existing != nil {
		return Fail[*User](ErrAlreadyExists, "EMAIL_EXISTS")
	}

	// Check if username exists
	existing, _ = userRepo.FindByUsername(ctx, username)
	if existing != nil {
		return Fail[*User](ErrAlreadyExists, "USERNAME_EXISTS")
	}

	// Hash password
	hash, err := s.passwordHasher.Hash(password)
	if err != nil {
		return Fail[*User](err, "HASH_FAILED")
	}

	// Create user
	user := &User{
		Email:        email,
		Username:     username,
		PasswordHash: hash,
	}

	if err := s.repo.Create(ctx, user); err != nil {
		return Fail[*User](err, "CREATE_FAILED")
	}

	// Publish event
	s.eventBus.Publish(ctx, Event{
		Type:        "user.registered",
		AggregateID: user.ID,
		Payload:     map[string]string{"email": email},
		OccurredAt:  time.Now(),
	})

	return Ok(user)
}

func (s *UserService) Authenticate(ctx context.Context, email, password string) Result[*User] {
	userRepo := s.repo.(UserRepository)

	user, err := userRepo.FindByEmail(ctx, email)
	if err != nil || user == nil {
		return Fail[*User](ErrUnauthorized, "INVALID_CREDENTIALS")
	}

	if !s.passwordHasher.Verify(password, user.PasswordHash) {
		return Fail[*User](ErrUnauthorized, "INVALID_CREDENTIALS")
	}

	userRepo.UpdateLastLogin(ctx, user.ID)

	s.eventBus.Publish(ctx, Event{
		Type:        "user.authenticated",
		AggregateID: user.ID,
		Payload:     map[string]string{"email": email},
		OccurredAt:  time.Now(),
	})

	return Ok(user)
}

File: _complete_archive/_archive_2026-02-26_18-50-32/stacks/node/base/code/service_layer.tpl.js
/*
File: service_layer.tpl.js
Purpose: Service layer pattern with dependency injection
Generated for: {{PROJECT_NAME}}
*/

// Result pattern for service operations
class Result {
    constructor(success, data = null, error = null, errorCode = null) {
        this.success = success;
        this.data = data;
        this.error = error;
        this.errorCode = errorCode;
    }

    static ok(data) {
        return new Result(true, data);
    }

    static fail(error, errorCode = 'ERROR') {
        return new Result(false, null, error, errorCode);
    }

    isOk() {
        return this.success;
    }

    isFail() {
        return !this.success;
    }
}

// Base service class
class BaseService {
    constructor(repository) {
        this.repository = repository;
    }

    async get(id) {
        const entity = await this.repository.findById(id);
        if (!entity) {
            return Result.fail('Entity not found', 'NOT_FOUND');
        }
        return Result.ok(entity);
    }

    async list(options = {}) {
        const result = await this.repository.findAll(options);
        return Result.ok(result);
    }

    async create(data) {
        try {
            const entity = await this.repository.create(data);
            return Result.ok(entity);
        } catch (error) {
            return Result.fail(error.message, 'CREATE_FAILED');
        }
    }

    async update(id, data) {
        try {
            const entity = await this.repository.update(id, data);
            return Result.ok(entity);
        } catch (error) {
            return Result.fail(error.message, 'UPDATE_FAILED');
        }
    }

    async delete(id) {
        try {
            await this.repository.delete(id);
            return Result.ok({ deleted: true });
        } catch (error) {
            return Result.fail(error.message, 'DELETE_FAILED');
        }
    }
}

// Event emitter for domain events
const { EventEmitter } = require('events');

class EventBus extends EventEmitter {
    constructor() {
        super();
        this.setMaxListeners(100);
    }

    async publish(eventType, payload) {
        this.emit(eventType, {
            type: eventType,
            payload,
            occurredAt: new Date(),
        });
    }

    subscribe(eventType, handler) {
        this.on(eventType, handler);
        return () => this.off(eventType, handler);
    }
}

// Simple dependency injection container
class Container {
    constructor() {
        this.services = new Map();
        this.factories = new Map();
        this.singletons = new Map();
    }

    register(name, instance) {
        this.services.set(name, instance);
        return this;
    }

    registerFactory(name, factory, singleton = true) {
        this.factories.set(name, { factory, singleton });
        return this;
    }

    resolve(name) {
        // Check if already instantiated
        if (this.services.has(name)) {
            return this.services.get(name);
        }

        // Check singletons
        if (this.singletons.has(name)) {
            return this.singletons.get(name);
        }

        // Check factories
        if (this.factories.has(name)) {
            const { factory, singleton } = this.factories.get(name);
            const instance = factory(this);

            if (singleton) {
                this.singletons.set(name, instance);
            }

            return instance;
        }

        throw new Error(`Service '${name}' not registered`);
    }

    has(name) {
        return this.services.has(name) || this.factories.has(name) || this.singletons.has(name);
    }
}

// Example user service
class UserService extends BaseService {
    constructor(repository, eventBus, passwordService) {
        super(repository);
        this.eventBus = eventBus;
        this.passwordService = passwordService;
    }

    async register(email, password, username) {
        // Check if email exists
        const existingEmail = await this.repository.findByEmail(email);
        if (existingEmail) {
            return Result.fail('Email already registered', 'EMAIL_EXISTS');
        }

        // Check if username exists
        const existingUsername = await this.repository.findByUsername(username);
        if (existingUsername) {
            return Result.fail('Username already taken', 'USERNAME_EXISTS');
        }

        // Hash password
        const passwordHash = await this.passwordService.hash(password);

        // Create user
        const user = await this.repository.create({
            email,
            username,
            passwordHash,
        });

        // Publish event
        await this.eventBus.publish('user.registered', {
            userId: user.id,
            email: user.email,
        });

        return Result.ok(user);
    }

    async authenticate(email, password) {
        const user = await this.repository.findByEmail(email);
        if (!user) {
            return Result.fail('Invalid credentials', 'INVALID_CREDENTIALS');
        }

        const isValid = await this.passwordService.verify(password, user.passwordHash);
        if (!isValid) {
            return Result.fail('Invalid credentials', 'INVALID_CREDENTIALS');
        }

        await this.repository.updateLastLogin(user.id);

        await this.eventBus.publish('user.authenticated', {
            userId: user.id,
            email: user.email,
        });

        return Result.ok(user);
    }

    async changePassword(userId, currentPassword, newPassword) {
        const result = await this.get(userId);
        if (result.isFail()) return result;

        const user = result.data;
        const isValid = await this.passwordService.verify(currentPassword, user.passwordHash);
        if (!isValid) {
            return Result.fail('Current password is incorrect', 'INVALID_PASSWORD');
        }

        const newHash = await this.passwordService.hash(newPassword);
        await this.repository.update(userId, { passwordHash: newHash });

        await this.eventBus.publish('user.password_changed', { userId });

        return Result.ok({ changed: true });
    }
}

// Password service
class PasswordService {
    constructor(bcrypt) {
        this.bcrypt = bcrypt;
        this.rounds = 12;
    }

    async hash(password) {
        return this.bcrypt.hash(password, this.rounds);
    }

    async verify(password, hash) {
        return this.bcrypt.compare(password, hash);
    }
}

// Setup container
function setupContainer(dependencies = {}) {
    const container = new Container();

    // Register event bus
    container.register('eventBus', new EventBus());

    // Register bcrypt
    container.register('bcrypt', dependencies.bcrypt || require('bcrypt'));

    // Register password service
    container.registerFactory('passwordService', (c) => {
        return new PasswordService(c.resolve('bcrypt'));
    });

    // Register repositories (pass in from dependencies)
    if (dependencies.userRepository) {
        container.register('userRepository', dependencies.userRepository);
    }

    // Register services
    container.registerFactory('userService', (c) => {
        return new UserService(
            c.resolve('userRepository'),
            c.resolve('eventBus'),
            c.resolve('passwordService')
        );
    });

    return container;
}

module.exports = {
    Result,
    BaseService,
    EventBus,
    Container,
    UserService,
    PasswordService,
    setupContainer,
};
